python39Packages.pybind11 fails to cross-compile

### Steps To Reproduce
Steps to reproduce the behavior:
1. In current "master" (c7186043a03336a249945fea67d41932bb648541)
2. `nix build .#pkgsCross.aarch64-multiplatform.python39Packages.pybind11`

### Build log
```
$ nix build  .#pkgsCross.aarch64-multiplatform.python39Packages.pybind11
error: assertion '(enablePython -> ((stdenv).hostPlatform == (stdenv).buildPlatform))' failed

       at /nix/store/7fq48sakdk0zj14pabfqaxlrpr6wcawm-source/pkgs/development/libraries/boost/generic.nix:32:1:

           31| # Python isn't supported when cross-compiling
           32| assert enablePython -> stdenv.hostPlatform == stdenv.buildPlatform;
             | ^
           33| assert enableNumpy -> enablePython;
(use '--show-trace' to show detailed location information)
```

When removing the check in boost (which seems invalid, it does build fine):
```
$ nix build  .#pkgsCross.aarch64-multiplatform.python39Packages.pybind11
warning: Git tree '/home/yvesf/vcs/nixpkgs' is dirty
error: builder for '/nix/store/vi22rql82c61cqjn4byd80ryckgaabjz-python3.9-pybind11-2.10.0-aarch64-unknown-linux-gnu.drv' failed with exit code 1;
       last 10 log lines:
       >   line 2: syntax error: unexpected word (expecting ")")
       >
       > Call Stack (most recent call first):
       >   tools/pybind11Tools.cmake:50 (find_package)
       >   tools/pybind11Common.cmake:180 (include)
       >   CMakeLists.txt:200 (include)
       >
       >
       > -- Configuring incomplete, errors occurred!
       > See also "/build/source/CMakeFiles/CMakeOutput.log".
       For full logs, run 'nix log /nix/store/vi22rql82c61cqjn4byd80ryckgaabjz-python3.9-pybind11-2.10.0-aarch64-unknown-linux-gnu.drv'.
```
actual reason is:
```
-- Found PythonInterp: /nix/store/kvflziwsfisrfqpjmwx7z0z15ykg77fj-python3-aarch64-unknown-linux-gnu-3.9.13/bin/python>
CMake Error at tools/FindPythonLibsNew.cmake:147 (message):
  Python config failure:


  /nix/store/kvflziwsfisrfqpjmwx7z0z15ykg77fj-python3-aarch64-unknown-linux-gnu-3.9.13/bin/python3:
  line 1: ^?ELF^B^A^A^B<B7>^A: not found
```

It tries to use the aarch64 python3.9 executable on the host system (x84_64).


### Metadata
```
nix run sys#nix-info -- -m
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.132, NixOS, 22.05 (Quokka), 22.05.20220728.9370544`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.8.1`
 - channels(yvesf): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

--- a/pkgs/development/libraries/boost/generic.nix
+++ b/pkgs/development/libraries/boost/generic.nix
@@ -28,8 +28,6 @@
 # We must build at least one type of libraries
 assert enableShared || enableStatic;
 
-# Python isn't supported when cross-compiling
-assert enablePython -> stdenv.hostPlatform == stdenv.buildPlatform;
 assert enableNumpy -> enablePython;
 
 # Boost <1.69 can't be built on linux with clang >8, because pth was removed

--- a/pkgs/development/python-modules/pybind11/default.nix
+++ b/pkgs/development/python-modules/pybind11/default.nix
@@ -34,6 +34,7 @@ buildPythonPackage rec {
     "-DBoost_INCLUDE_DIR=${lib.getDev boost}/include"
     "-DEIGEN3_INCLUDE_DIR=${lib.getDev eigen}/include/eigen3"
     "-DBUILD_TESTING=on"
+    "-DPYTHON_EXECUTABLE:FILEPATH=${python.pythonForBuild.interpreter}"
   ] ++ lib.optionals (python.isPy3k && !stdenv.cc.isClang) [
     "-DPYBIND11_CXX_STANDARD=-std=c++17"
   ];

